osu.Game.Rulesets.Taiko/Mods/TaikoModClassic.cs

```diff
@@ -12,16 +17,98 @@ namespace osu.Game.Rulesets.Taiko.Mods
 {
     public class TaikoModClassic : ModClassic, IApplicableToDrawableRuleset<TaikoHitObject>, IApplicableToDrawableHitObject
     {
+        private IReadOnlyList<Mod> mods = Array.Empty<Mod>();
+
+        private const float hd_base_fade_out_duration = 0.375f;
+
+        private const float hd_base_initial_alpha = 0.75f;
+
+        private const float hdhr_base_fade_out_duration = 0.2f;
+
+        private const float hdhr_base_initial_alpha = 0.2f;
+
+        private const float hidden_base_aspect = 4f / 3f;
+
+        private readonly BindableFloat hiddenFadeOutDuration = new BindableFloat(hd_base_fade_out_duration);
+
+        private readonly BindableFloat hiddenInitialAlpha = new BindableFloat(hd_base_initial_alpha);
+
         public void ApplyToDrawableRuleset(DrawableRuleset<TaikoHitObject> drawableRuleset)
         {
             var drawableTaikoRuleset = (DrawableTaikoRuleset)drawableRuleset;
-            drawableTaikoRuleset.LockPlayfieldAspectRange.Value = false;
+
+            // drawableRuleset.Mods should always be non-null here, but just in case.
+            mods = drawableRuleset.Mods ?? mods;
+
+            drawableTaikoRuleset.MaximumAspect.Value = 22f / 9f;
```

Comment:
This is to match stable with classic mod, stable allows a wider aspect ratio range

Reply:
the notes fade in at some point, and while you can't see the fade in with 16:9, you can see it with a higher aspect ratio

Reply:
I mean sure but that's not evidence of there being a aspect ratio lock on stable. That is also consistent with _there not being any lock in stable_, isn't it? Which [we had set already but this PR removes?](https://github.com/ppy/osu/pull/27136#discussion_r1487359132)

Reply:
I don't have access to stable's code. But I'm guessing for stable, while there isn't an explicit aspect ratio upper bound for the playfield itself, it still effectively limits the time rage to a ~2:1 aspect ratio equivalent by hiding notes beyond that point (fading them in. Without this high sv will lose most of their difficulty since player can in theory stretch the window and have indefinitely long time range.

> if you go wider than 16:9 more is shown

For stable nm, it seems to start fading notes in beyond ~2:1. 16:9 doesn't seem to be a cutoff point for stable nm.

With this PR it is done by trimming the playfield instead. This way we can reuse the trim mechanism used by classic HD, while keeping the upper bound of time range.

Reply:
> it still effectively limits the time rage to a ~2:1 aspect ratio equivalent by hiding notes beyond that point (fading them in

Now we're getting somewhere...

I have experimentally confirmed this on a single taiko map. Graphs are here: https://www.desmos.com/calculator/cdmlh4lm3u

But that generally makes me nervous as to whether it's valid to simulate the limit by clamping aspect ratio in the first place. Clamping ratio is a hard cutoff, it doesn't emulate limited visibility of the objects as they move onto the playfield, so I don't know that this should be done? Maybe the objects should just actually be fading in instead to match stable.

@ppy/team-client would appreciate thoughts on this spectacle if you can because I'm feeling increasingly out of my depth here in making shot calls as to how this should behave

Reply:
> this means that a fast-moving SV-accelerated circle is going to spend more of its active time fading in on screen than a normal circle would

Never really noticed this previously when playing normally, but I can confirm that this is true. When SV on stable nomod is high enough, it will still be fading in even with 16:9 aspect ratio

With this insight it does seem better to actually implement the fade in (trim was used under the assumption that notes take a fixed amount of space to fade in, which has proven to be false)

Reply:
Don't have access to my local branch right now, but fade in has been implemented. I'll be pushing it & posting comparison videos either tonight or some time tomorrow